i6_models/primitives/mixup.py

```diff
@@ -0,0 +1,153 @@
+        max_num_mix: maximum number of mixups (random int in [1, max_num_mix])
+    """
+
+    buffer_size: int = 1_000_000
+    apply_prob: float = 1.0
+    max_num_mix: int = 4
+    lambda_min: float = 0.003
+    lambda_max: float = 0.3
+
+
+class FeatureBuffer:
+    """
+    The FeatureBuffer saves the feature from previous timeframes
+    after the buffer is full, the buffer will not be updated
+    """
+
+    def __init__(self, buffer_size: int, feature_dim: int):
+        self.filled = False
+        self.pos = 0
+        self.buffer_size = buffer_size
+        self.cache = torch.nn.parameter.Parameter(
+            data=torch.zeros((self.buffer_size, feature_dim)), requires_grad=False
+        )
+
+    def append(self, tensor: torch.Tensor):
+        t_dim = tensor.shape[0]
+        if t_dim > self.buffer_size:
+            tensor = tensor[: self.buffer_size]
+            t_dim = self.buffer_size
+        delta_pos = min(self.buffer_size - self.pos, t_dim)
+        end_pos = self.pos + delta_pos
+        self.cache[self.pos : end_pos] = tensor[:delta_pos]
+        self.pos = end_pos
+
+        if end_pos == self.buffer_size:
+            self.filled = True
+            end_pos = t_dim - delta_pos
+            self.cache[:end_pos] = tensor[delta_pos:]
+            self.pos = end_pos
+
+    def getRandom(self, b_dim: int, t_dim: int, max_num_mixup: int, n_mask: torch.tensor) -> torch.Tensor:
+        if not self.filled and self.pos == 0:
+            return None
+        else:
+            end_idx = self.buffer_size if self.filled else self.pos
+            max_end_idx = end_idx - t_dim
+
+            start_indicies = torch.randint(
+                high=max_end_idx, size=(b_dim, max_num_mixup)
+            )  # [B, M] (M denotes maximum of num_mixup over the batch)
+            start_indicies_flat = torch.masked_select(
+                start_indicies, n_mask
+            )  # [B, M'] (M' denotes sum of num_mixup over the batch)
+
+            idx = torch.arange(t_dim)
+            idx = torch.unsqueeze(idx, dim=-1) + start_indicies_flat  # [T, M']
+            mixup_values = self.cache[idx]  # [T, M', F]
+            return mixup_values
+
+
+class Mixup:
+    """
+    Implement the Mixup data augmentation method
+    C.f. https://arxiv.org/abs/1710.09412
+
+    The code is partly based on Albert's implementation in returnn front-end
+    C.f. https://github.com/rwth-i6/i6_experiments/blob/main/users/zeyer/returnn/models/rf_mixup.py
+    """
+
+    def __init__(self, feature_dim: int, cfg: MixupConfig):
+        self.apply_prob = cfg.apply_prob
+        self.lambda_min = cfg.lambda_min
+        self.lambda_max = cfg.lambda_max
+        self.max_num_mix = cfg.max_num_mix
+        self.buffer_size = cfg.buffer_size
+        self.feature_dim = feature_dim
+
+        self.feature_buffer = FeatureBuffer(self.buffer_size, feature_dim)
+
+    def __call__(self, input: torch.Tensor, sequence_mask: torch.Tensor) -> torch.Tensor:
```

Comment:
If you make `Mixup` a proper Torch module (I think it should be), then here:
```suggestion
    def forward(self, input: torch.Tensor, *, sequence_mask: torch.Tensor) -> torch.Tensor:
```